IdentityCore/src/workplacejoin/MSIDWorkPlaceJoinUtilBase.m

```diff
@@ -173,4 +203,73 @@ For asymmetric keys this holds the public key hash which allows digital identity
     return keyPair;
 }
 
++ (MSIDWPJKeyPairWithCert *)getWPJKeysWithTenantId:(__unused NSString *)tenantId context:(__unused id<MSIDRequestContext>)context
+{
+    NSString *teamId = [[MSIDKeychainUtil sharedInstance] teamId];
+    
+    if (!teamId)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Encountered an error when reading teamID from keychain.");
+        return nil;
+    }
+    
+    NSString *legacySharedAccessGroup = [NSString stringWithFormat:@"%@.com.microsoft.workplacejoin", teamId];
+    NSData *tagData = [kMSIDPrivateKeyIdentifier dataUsingEncoding:NSUTF8StringEncoding];
+    
+    NSDictionary *extraPrivateKeyAttributes = @{ (__bridge id)kSecAttrApplicationTag: tagData,
+                                                 (__bridge id)kSecAttrAccessGroup : legacySharedAccessGroup };
+    NSDictionary *extraCertAttributes = @{ (__bridge id)kSecAttrAccessGroup : legacySharedAccessGroup };
+    
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Checking Legacy keychain for registration.");
+    MSIDWPJKeyPairWithCert *legacyKeys = [self findWPJRegistrationInfoWithAdditionalPrivateKeyAttributes:extraPrivateKeyAttributes certAttributes:extraCertAttributes context:context shouldCheckEnclave:NO];
+        
+    if (legacyKeys)
+    {
+        if ([NSString msidIsStringNilOrBlank:tenantId])
+        {
+            // ESTS didn't request a specific tenant, just return default one
+            return legacyKeys;
+        }
+        
+        // Read tenantId for legacy identity
+        NSError *tenantIdError = nil;
+        NSString *registrationTenantId = [MSIDWorkPlaceJoinUtil getWPJStringDataForIdentifier:kMSIDTenantKeyIdentifier context:context error:&tenantIdError];
+        
+        // There's no tenantId on the registration, or it mismatches what server requested, keep looking for a better match. Otherwise, return the identity already.
+        if (!tenantIdError
+            && registrationTenantId
+            && [registrationTenantId isEqualToString:tenantId])
+        {
+            return legacyKeys;
+        }
+    }
+    
+    if (!tenantId)
+    {
+        // default registration should have a tenantId associated.
+        return legacyKeys;
+    }
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Checking Default keychain for registration.");
+    NSString *defaultSharedAccessGroup = [NSString stringWithFormat:@"%@.com.microsoft.workplacejoin.v2", teamId];
```

Comment:
I think that is right way to query new private key (default RSA, ECC) that will be stored in the iCloud keychain both iOS/MacOS.
Just curious, for legacy privatekey, on iOS it's still in iCloud keychain, on MacOS it's ACL so I guess should be 2 diff queries. However I don't see you prepare 2 queries based on diff platform. Might be I missed somewhere?
The same question for certificate: on MacOS default RSA cert stored still in login keychain (ACL). ECC cert stored in iCloud. So I guess we need 2 queries here but didn't see that.
I will loop more at the query in WPJ later, maybe I forget some detail there.

Reply:
Per our offline discussions, Legacy private key/certificate is not stored using ACL but metadata is.